mythril_core/src/acpi.rs

```diff
@@ -0,0 +1,201 @@
+    },
+    /// RSDP structure variant for version 2 (`revision == 2`).
+    V2 {
+        /// OEM Supplied string.
+        oemid: [u8; 6],
+        /// 32-bit physical address of the RSDT.
+        rsdt_addr: u32,
+        /// Length of the XSDT table in bytes.
+        length: u32,
+        /// 64-bit physical address of the XSDT.
+        xsdt_addr: u64,
+    },
+}
+
+impl fmt::Debug for RSDP {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &RSDP::V1 { rsdt_addr, .. } => {
+                write!(f, "RSDP 1.0: rsdt=0x{:x}", rsdt_addr)
+            }
+            &RSDP::V2 {
+                rsdt_addr,
+                xsdt_addr,
+                ..
+            } => write!(
+                f,
+                "RSDP 2.0: rsdt=0x{:x} xsdt_addr=0x{:x}",
+                rsdt_addr, xsdt_addr
+            ),
+        }
+    }
+}
+
+impl RSDP {
+    /// Find the RSDP in the Extended BIOS Data Area or the
+    /// Main BIOS area.
+    ///
+    /// This is described in [ACPI § 5.2.5.1]
+    ///
+    /// [ACPI § 5.2.5.1]: https://uefi.org/sites/default/files/resources/ACPI_6_3_May16.pdf
+    pub fn find() -> Result<RSDP> {
+        let bytes = unsafe {
+            // Try to find the RSDP in the Main BIOS area.
+            Self::search_range(EXTENDED_BIOS_DATA_START, EXTENDED_BIOS_DATA_END)
+                .or_else(|_| {
+                    // If we didn't find the RSDP in the Main BIOS area, try
+                    // the Extended BIOS Area
+                    Self::search_range(MAIN_BIOS_DATA_START, MAIN_BIOS_DATA_END)
+                })?
+        };
+
+        // Extract the OEMID, revision, and RSDT address from the address
+        // found regardless of the ACPI version.
+        let mut oemid = [0u8; offsets::OEMID.end - offsets::OEMID.start];
+        oemid.copy_from_slice(&bytes[offsets::OEMID]);
+
+        let rsdt_addr = NativeEndian::read_u32(&bytes[offsets::RSDT_ADDR]);
+
+        let rsdp = match bytes[offsets::REVISION] {
+            // If this is ACPI 1.0, there is no XSDT.
+            0 => RSDP::V1 { oemid, rsdt_addr },
+            // This is ACPI 2.0. Extract the addrss and length of the XSDT.
+            2 => RSDP::V2 {
+                oemid,
+                rsdt_addr,
+                length: NativeEndian::read_u32(&bytes[offsets::LENGTH]),
+                xsdt_addr: NativeEndian::read_u64(&bytes[offsets::XSDT_ADDR]),
+            },
+            _ => {
+                return Err(Error::InvalidValue(format!(
+                    "Invalid RSDP revision: {}",
+                    bytes[offsets::REVISION]
+                )))
+                .into();
+            }
+        };
+
+        // Now that we know this is a version we support, validate the checksum.
+        Self::verify_rsdp_checksum(bytes)?;
+        Ok(rsdp)
+    }
+
+    /// Search a given address range for the RSDP.
+    unsafe fn search_range(start: usize, end: usize) -> Result<&'static [u8]> {
```

Comment:
Sorry to keep adding suggestions, but there's something I don't really like about this signature. Maybe we could `slice::from_raw_parts` outside this function, then the lifetime of the returned bytes are bound to the lifetime of the input, rather than static. I think it might also help avoid some of the pointer magic going on in here.

Reply:
Np, good suggestion. Updated to create a slice and I do think it looks better now.